sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,829 @@
+}
+
+#[cfg(not(target_arch = "bpf"))]
+#[derive(AbiExample, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct BigNumber {
+    data: openssl::bn::BigNum,
+}
+#[cfg(target_arch = "bpf")]
+#[derive(AbiExample, Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct BigNumber {
+    data: Vec<u8>,
+    is_negative: bool,
+}
+
+impl BigNumber {
+    /// Returns a BigNumber with initial value of 0
+    pub fn new() -> Self {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            let bn = openssl::bn::BigNum::new().unwrap();
+            Self { data: bn }
+        }
+        #[cfg(target_arch = "bpf")]
+        Self {
+            data: vec![0],
+            is_negative: false,
+        }
+    }
+
+    /// Returns the size, in bytes, of BigNum. Typically used in
+    /// assessing buffer size needed to fetch the bytes for serialization, etc.
+    pub fn size_in_bytes(&self) -> usize {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            self.data.num_bytes() as usize
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            self.data.len()
+        }
+    }
+
+    /// Returns a big-endian byte vector representation
+    /// of the absolute value of `self`
+    pub fn to_vec(&self) -> Vec<u8> {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            if self.data.num_bytes() == 0 {
+                return vec![0u8];
+            }
+            self.data.to_vec()
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            self.data.clone()
+        }
+    }
+
+    #[cfg(target_arch = "bpf")]
+    pub fn as_ref(&self) -> &[u8] {
+        &self.data
+    }
+
+    /// Returns a BigNumber with initial value set to a u32 value
+    pub fn from_u32(u32_val: u32) -> Self {
+        if u32_val == 0 {
+            Self::new()
+        } else {
+            #[cfg(not(target_arch = "bpf"))]
+            {
+                let bn = openssl::bn::BigNum::from_u32(u32_val).unwrap();
+                Self { data: bn }
+            }
+            #[cfg(target_arch = "bpf")]
+            {
+                extern "C" {
+                    fn sol_bignum_from_u32(bn_syscall_result: *mut u64, u32_data: u64) -> u64;
+                }
+                let mut value = Vec::<u8>::with_capacity(4);
+                let mut bn_syscall_result = {
+                    FfiBigNumber {
+                        data: value.as_mut_ptr() as *mut _ as u64,
+                        data_len: 4u64,
+                        is_negative: false,
+                    }
+                };
+                unsafe {
+                    sol_bignum_from_u32(
+                        &mut bn_syscall_result as *mut _ as *mut u64,
+                        u32_val as u64,
+                    );
+                    value.set_len(bn_syscall_result.data_len as usize);
+                }
+                Self {
+                    is_negative: bn_syscall_result.is_negative,
+                    data: value,
```

Comment:
So this can be written as:

```
let mut value = u32_val.to_be_bytes().to_vec();
while value.len() > 1 && value[0] == 0 {
    value.remove(0);
}
```
Do we need a syscall for that?

Reply:
Commented previously, this gives us cost control.

Reply:
@seanyoung One thing this does is removes any assumption the program has to make on the encoding of `value` which could be important later if the underlying implementation changes

Reply:
Folks are going to store these encoded numbers in account data so we are probably locked in with this particular encoding for this api

Reply:
Strike that, encoded values will be recorded to account data by developers